packages/story-editor/src/app/highlights/quickActions/useQuickActions.js

```diff
@@ -94,6 +106,144 @@ const useQuickActions = () => {
+          optimizeGif,
+        };
+      }
+    );
+
+  const { isTranscodingEnabled } = useFFmpeg();
+
+  // Media Upload Props
+  const allowedMimeTypes = useMemo(() => {
+    if (isTranscodingEnabled) {
+      return [
+        ...allowedTranscodableMimeTypes,
+        ...allowedImageMimeTypes,
+        ...allowedVideoMimeTypes,
+      ];
+    }
+    return [...allowedImageMimeTypes, ...allowedVideoMimeTypes];
+  }, [
+    allowedImageMimeTypes,
+    allowedVideoMimeTypes,
+    isTranscodingEnabled,
+    allowedTranscodableMimeTypes,
+  ]);
+
+  const transcodableMimeTypes = useMemo(() => {
+    return allowedTranscodableMimeTypes.filter(
+      (x) => !allowedVideoMimeTypes.includes(x)
+    );
+  }, [allowedTranscodableMimeTypes, allowedVideoMimeTypes]);
+
+  let onSelectErrorMessage = __(
+    'No file types are currently supported.',
+    'web-stories'
+  );
+  if (allowedFileTypes.length) {
+    onSelectErrorMessage = sprintf(
+      /* translators: %s: list of allowed file types. */
+      __('Please choose only %s to insert into page.', 'web-stories'),
+      translateToExclusiveList(allowedFileTypes)
+    );
+  }
+
+  /**
+   * Insert element such image, video and audio into the editor.
+   *
+   * @param {Object} resource Resource object
+   * @param {string} thumbnailURL The thumbnail's url
+   * @return {null|*} Return onInsert or null.
+   */
+  const insertMediaElement = useCallback(
+    (resource, thumbnailURL) => {
+      resourceList.set(resource.id, {
+        url: thumbnailURL,
+        type: 'cached',
+      });
+      updateElementsById({
+        elementIds: [selectedElements?.[0]?.id],
+        properties: { type: resource.type, resource },
+      });
+    },
+    [selectedElements, updateElementsById]
+  );
+
+  const handleMediaSelect = useCallback(
+    (resource) => {
+      try {
+        if (isTranscodingEnabled && canTranscodeResource(resource)) {
+          if (transcodableMimeTypes.includes(resource.mimeType)) {
+            optimizeVideo({ resource });
+          }
+
+          if (resource.mimeType === 'image/gif') {
+            optimizeGif({ resource });
+          }
+        }
+        // WordPress media picker event, sizes.medium.source_url is the smallest image
+        insertMediaElement(
+          resource,
+          resource.sizes?.medium?.source_url || resource.src
+        );
+
+        if (
+          !resource.local &&
+          allowedVideoMimeTypes.includes(resource.mimeType) &&
+          resource.isMuted === null
+        ) {
+          updateVideoIsMuted(resource.id, resource.src);
+        }
+      } catch (e) {
+        showSnackbar({
+          message: e.message,
+          dismissable: true,
+        });
+      }
+    },
+    [
+      allowedVideoMimeTypes,
+      insertMediaElement,
+      isTranscodingEnabled,
+      optimizeGif,
+      optimizeVideo,
+      showSnackbar,
+      transcodableMimeTypes,
+      updateVideoIsMuted,
+    ]
+  );
+
+  const MediaPicker = useCallback(
```

Comment:
Even though we are using `useCallback`, isn't it odd to define a new component inside a custom hook? Is it possible to move it out?

Reply:
@choumx wanted it to be inlined here in the same hook: https://github.com/google/web-stories-wp/pull/9534#discussion_r743952018

We can leave it here or move it outside the hook. Either is fine with me.

Reply:
As this has passed QA, thoughts on keeping as is, inline with @choumx 's request? @sayedtaqui and @swissspidy do you want this moved out or do you want to leave it as is? This PR's been open for a long time, it'd be nice to have it done.

Reply:
I think it should be in the same file, either inside or outside of the hook is fine. IMO it's ok as-is.

Reply:
I'd keep it in the same file but outside the hook if possible.